drivers/can/can_stm32.c

```diff
@@ -1126,47 +1139,52 @@ NET_DEVICE_INIT(socket_can_stm32_1, SOCKET_CAN_NAME_1, socket_can_init_1,
 
 #endif /* CONFIG_NET_SOCKETS_CAN */
 
-#endif /*CONFIG_CAN_1*/
+#endif /* DT_HAS_NODE(DT_NODELABEL(can1)) */
 
-#ifdef CONFIG_CAN_2
+#if DT_HAS_NODE(DT_NODELABEL(can2)) && \
+	DT_NODE_HAS_COMPAT(DT_NODELABEL(can2), st_stm32_can)
 
 static void config_can_2_irq(CAN_TypeDef *can);
 
 static const struct can_stm32_config can_stm32_cfg_2 = {
-	.can = (CAN_TypeDef *)DT_CAN_2_BASE_ADDRESS,
-	.master_can = (CAN_TypeDef *)DT_CAN_1_BASE_ADDRESS,
-	.bus_speed = DT_CAN_2_BUS_SPEED,
-	.sjw = DT_CAN_2_SJW,
-	.prop_ts1 = DT_CAN_2_PROP_SEG + DT_CAN_2_PHASE_SEG1,
-	.ts2 = DT_CAN_2_PHASE_SEG2,
+	.can = (CAN_TypeDef *)DT_REG_ADDR(DT_NODELABEL(can2)),
+	.master_can = (CAN_TypeDef *)DT_REG_ADDR(DT_NODELABEL(can1)),
```

Comment:
@alexanderwachter, @galak this is the main impact of the change.
To use dma2, we now need to compile dma1.
To handle this, I added 3e77fe090828b2b760ad509aad702c897fc03c56

Reply:
That's no the only impact. We are also wasting memory because we instantiate both deriver instances. The reason why we have the CAN_2 is because of pin multiplexing. But on Slack somebody announced to implement support for both CAN simultaneously. Maybe we can find another solution to reference to CAN_1. Maybe with a phandle?

Reply:
> We are also wasting memory because we instantiate both driver instances.

Of course

> Maybe we can find another solution to reference to CAN_1. Maybe with a phandle?

Maybe it can be done through dedicated prop on can2 description, something like master-can, or whatever. This would be optional as can1 would not need it. Can you check this particular point?

Reply:
Edited PR description on various attempts to solve this.